### DISCUSSION TITLE
rm: cannot remove '/etc/nginx/conf.d/*': No such file or directory

### DISCUSSION THREAD

**Discussion by IavTavares** [nginx/kubernetes-ingress#2522] [Q&A] [answered]
I'm trying to dockerize  a flask + gunicorn app with nginx.

Here's the initial part of my dockerfile:

```
FROM ubuntu
ARG DEBIAN_FRONTEND=noninteractive
ARG DEBCONF_NONINTERACTIVE_SEEN=true
# it seems to be important ARGs related to shell come before any run
RUN apt-get update && apt-get install -y python3 python3-pip ffmpeg libsm6 libxext6
RUN apt-get install -y imagemagick

RUN apt-get install nginx -y

COPY . . 
RUN rm /etc/nginx/conf.d/default.conf
```
when building the docker image, I get the error message: `rm: cannot remove '/etc/nginx/conf.d/default.conf': No such file or directory`.

Even if i change the RUN command to `RUN rm /etc/nginx/conf.d/*`, I still get the same error message. It makes me think that there's no directory with that name.

But isn't it supposed to exit according to the documentation?

**Comment by brianehlert** [comment]
you might find this blog useful: https://www.nginx.com/blog/deploying-nginx-nginx-plus-docker/
Also, run your container and verify the path.  I believe you are expecting a very old path.

**Reply by lucacome** [reply]
@IavTavares the version of the `nginx` package you are installing with `apt` is the one maintained by OS (in this case ubuntu) and it has different default files compared to the one packaged by NGINX. 
You can see the steps here to add the NGINX repository https://nginx.org/en/linux_packages.html#Ubuntu and install the package maintained by NGINX. 

Let me know if it makes sense.

**Reply by IavTavares** [reply]
@brianehlert Thanks for the reply. In the link you shared, there's this line of code for the dockerfile:
«RUN rm /etc/nginx/nginx.conf /etc/nginx/conf.d/default.conf»
Do you happen to know what is the new path?

**Reply by IavTavares** [reply]
@lucacome Thanks for the reply. I'm not sure I understand the link you shared... They only shows us how to make sure that the package is the right one, and then run the usual «sudo apt install nginx». It doesn't tell us what is the directory where one can expect to find the nginx configuration files.

Edit: On a second reading, I think I understand you a bit better. I should use the mainline nginx file list to be able to use the path in my dockerfile. Otherwise, using Ubuntu's package, we get different paths/files.

**Reply by lucacome** [reply]
yeah that's exactly right @IavTavares , if you use the mainline repository, you will see those files 🙂

**Reply by IavTavares** [reply]
@lucacome So, now, my dockerfile looks like this:
```
FROM ubuntu
ARG DEBIAN_FRONTEND=noninteractive
ARG DEBCONF_NONINTERACTIVE_SEEN=true

RUN apt-get update && apt-get install -y python3 python3-pip ffmpeg libsm6 libxext6
RUN apt-get install -y imagemagick

# choosing mainline nginx image
RUN echo "deb [signed-by=/usr/share/keyrings/nginx-archive-keyring.gpg] \
http://nginx.org/packages/mainline/ubuntu `lsb_release -cs` nginx" \
    | > /etc/apt/sources.list.d/nginx.list
RUN echo -e "Package: *\nPin: origin nginx.org\nPin: release o=nginx\nPin-Priority: 900\n" \
    | > /etc/apt/preferences.d/99nginx
# we need to update apt after this

RUN apt update && apt install nginx -y

COPY . . 
RUN rm /etc/nginx/conf.d/*
```
But I still get the same problem with the inexistent folder path...

Do you know why?

**Reply by lucacome** [reply]
@IavTavares there are a couple of problems:
- you didn't install the necessary dependencies, for example `lsb-release` so it won't be able to automatically resolve the ubuntu version (you could just replace `lsb_release -cs` with `focal` in this case or the version you're using)
- you're actually not adding anything any text to those files because you have a `|` before the `>`

```
FROM ubuntu
ARG DEBIAN_FRONTEND=noninteractive
ARG DEBCONF_NONINTERACTIVE_SEEN=true

RUN apt-get update && apt-get install -y python3 python3-pip ffmpeg libsm6 libxext6
RUN apt install -y curl gnupg2 ca-certificates lsb-release ubuntu-keyring
RUN apt-get install -y imagemagick

RUN curl -fssL https://nginx.org/keys/nginx_signing.key | gpg --dearmor > /usr/share/keyrings/nginx-archive-keyring.gpg

# choosing mainline nginx image
RUN printf "%s\n" "deb [signed-by=/usr/share/keyrings/nginx-archive-keyring.gpg] \
    http://nginx.org/packages/mainline/ubuntu `lsb_release -cs` nginx" > /etc/apt/sources.list.d/nginx.list
RUN printf "%s\n" "Package: *" "Pin: origin nginx.org" "Pin: release o=nginx" \
    "Pin-Priority: 900" > /etc/apt/preferences.d/99nginx
# we need to update apt after this

RUN apt update && apt install nginx -y

COPY . .
RUN rm /etc/nginx/conf.d/*
```
this is an example of how I changed your Dockerfile (I prefer to use `prinft` to `echo`, it's cleaner to me 😄 )

**Reply by IavTavares** [reply]
@lucacome Hi Luca, thanks for the help! Sorry for the late reply, but life happened... I'll try today your suggestion, and keep you posted. ;)

**Reply by IavTavares** [reply]
@lucacome 
So, Luca, I've tried the following:
```
RUN apt update && apt install -y python3 python3-pip ffmpeg libsm6 libxext6

RUN apt install -y curl gnupg2 ca-certificates lsb-release ubuntu-keyring

RUN curl -fssL https://nginx.org/keys/nginx_signing.key | gpg --dearmor > /usr/share/keyrings/nginx-archive-keyring.gpg

# choosing stable nginx image
RUN echo "deb [signed-by=/usr/share/keyrings/nginx-archive-keyring.gpg] \
    http://nginx.org/packages/ubuntu focal nginx" > /etc/apt/sources.list.d/nginx.list
RUN echo -e "Package: *\nPin: origin nginx.org\nPin: release o=nginx\nPin-Priority: 900\n" \
    > /etc/apt/preferences.d/99nginx

# we need to update apt after this
RUN apt update && apt install nginx -y
```
And got the following error message:
`E: Invalid record in the preferences file /etc/apt/preferences.d/99nginx, no Package header
`
I don't seem to find a solution online...

**Reply by lucacome** [reply]
@IavTavares 

it looks like `echo -e` is the problem as it's printing `-e` to the file. 
Docker uses `sh` as the default shell when building a Dockerfile (and in most distributions, `sh` is actually a symlink to `ash` that has `echo` as a builtin command and it doesn't support `-e`), so the implementation is different than for example in `bash` (I guess the most popular shell).

That's why I prefer to use `printf` 🙂 but you could also change the shell by adding `SHELL ["/bin/bash", "-c"]` in your Dockerfile, before the `RUN` commands

As a side note, usually when I see that a step is not behaving correctly or as I think it should, I just comment out the lines after that (in this case `RUN apt-update ...` and build the container, then I just run the container and inspect the files to see that everything is in order.

I hope it helps.

**Reply by IavTavares** [reply, marked as answer]
@lucacome  Many thanks for all the help. I've just solved the problem! ;)
